network/dag/bbolt_dag.go

```diff
@@ -227,42 +306,53 @@ func (dag *bboltDAG) Add(ctx context.Context, transactions ...Transaction) error
 	return nil
 }
 
-func (dag bboltDAG) Walk(ctx context.Context, algo WalkerAlgorithm, visitor Visitor, startAt hash.SHA256Hash) error {
+func (dag bboltDAG) Walk(ctx context.Context, visitor Visitor, startAt hash.SHA256Hash) error {
 	return bboltTXView(ctx, dag.db, func(ctx context.Context, tx *bbolt.Tx) error {
 		transactions := tx.Bucket([]byte(transactionsBucket))
-		nexts := tx.Bucket([]byte(nextsBucket))
-		if transactions == nil || nexts == nil {
+		clocksBucket := tx.Bucket([]byte(clockBucket))
+		if transactions == nil {
 			// DAG is empty
 			return nil
 		}
-		return algo.walk(ctx, visitor, startAt, func(hash hash.SHA256Hash) (Transaction, error) {
-			return getTransaction(hash, transactions)
-		}, func(hash hash.SHA256Hash) ([]hash.SHA256Hash, error) {
-			return parseHashList(nexts.Get(hash.Slice())), nil // no need to copy, calls FromSlice() (which copies)
-		})
-	})
-}
 
-func (dag bboltDAG) Root(ctx context.Context) (hash hash.SHA256Hash, err error) {
-	err = bboltTXView(ctx, dag.db, func(ctx context.Context, tx *bbolt.Tx) error {
-		if transactions := tx.Bucket([]byte(transactionsBucket)); transactions != nil {
-			if roots := getRoots(transactions); len(roots) >= 1 {
-				hash = roots[0]
+		// we find the clock value of the TX ref
+		// an empty hash means start at root
+		clockValue, err := getClock(tx, startAt)
```

Comment:
What if there are multiple TXs with the same clock value (due to a branch)? I'd expect it to only walk TXs from the given TX, not TXs (and subsequent TXs) from parallel branches with the same LC.

Reply:
it's true that a TX can be hit twice, although the `visitedTransactions` hash will block it. Also the callbacks should be resilient (and currently are) to duplicates. I added logic for checking if a TX is processed at this level, but it adds way too much lines of ugliness which is already caught at different points.
I also suspect the `startAt` will be replaced with clock value somewhere down the line anyway since it would simplify a lot.

Reply:
Given that:

- One of the important rules of thumb of the DAG is that we minimize branches (and have them as short as possible), and thus minimize the number of TXs that would be revisited.
- We specify in the contract of `Walk()` that more TXs than just the requested branch can be visited (e.g. if you want to build a visualization from TX `XYZ` you might also get parallel branch `ABC`).
- And thus make sure all consumers can handle consuming TXs multiple times (there was a good term for this, but I forgot).

I think it's probably safe for now.